// Node Plugin Service Protocol Buffer Definition
// Version: 2.0.0
// 
// 这个文件定义了节点插件的 gRPC 服务接口
// 设计原则：
// 1. 类型安全：使用结构化消息替代 JSON 字符串
// 2. 可扩展：预留字段号，使用 oneof 支持多种响应类型
// 3. 可追踪：包含执行上下文和追踪信息
// 4. 向后兼容：遵循 protobuf 最佳实践

syntax = "proto3";

package node_plugin;

option go_package = "github.com/gw123/gflow/plugins/base-go/proto";

// 注意：为简化跨语言兼容性，使用 int64 毫秒时间戳替代 google.protobuf.Timestamp

// =============================================================================
// 服务定义
// =============================================================================

// NodePluginService 定义节点插件的 gRPC 服务接口
service NodePluginService {
  // GetMetadata 获取插件元数据（插件能力描述）
  rpc GetMetadata(GetMetadataRequest) returns (GetMetadataResponse);
  
  // Init 初始化节点（在执行前调用）
  rpc Init(InitRequest) returns (InitResponse);
  
  // Run 执行节点（支持流式输出）
  rpc Run(RunRequest) returns (stream RunResponse);
  
  // Stop 停止正在执行的节点（优雅停止）
  rpc Stop(StopRequest) returns (StopResponse);
  
  // TestCredential 测试凭证有效性
  rpc TestCredential(TestCredentialRequest) returns (TestCredentialResponse);
  
  // HealthCheck 健康检查
  rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
}

// =============================================================================
// 通用类型定义
// =============================================================================

// Value 表示任意类型的值（类似于 google.protobuf.Value，但更精确）
message Value {
  oneof kind {
    NullValue null_value = 1;
    string string_value = 2;
    int64 int_value = 3;
    double double_value = 4;
    bool bool_value = 5;
    bytes bytes_value = 6;
    ListValue list_value = 7;
    MapValue map_value = 8;
  }
}

// NullValue 表示空值
enum NullValue {
  NULL_VALUE = 0;
}

// ListValue 表示数组
message ListValue {
  repeated Value values = 1;
}

// MapValue 表示对象/映射
message MapValue {
  map<string, Value> fields = 1;
}

// ExecutionContext 执行上下文信息
message ExecutionContext {
  string workflow_id = 1;       // 工作流 ID
  string execution_id = 2;      // 执行实例 ID
  string node_id = 3;           // 节点 ID
  string trace_id = 4;          // 追踪 ID（用于分布式追踪）
  string span_id = 5;           // Span ID
  int32 retry_count = 6;        // 重试次数（0 表示首次执行）
  int64 timeout_ms = 7;         // 超时时间（毫秒）
  map<string, string> metadata = 8;  // 额外元数据
}

// =============================================================================
// GetMetadata - 获取插件元数据
// =============================================================================

message GetMetadataRequest {
  // 预留字段供未来扩展
  reserved 1 to 10;
}

message GetMetadataResponse {
  // 基本信息
  string name = 1;                    // 插件名称（唯一标识）
  string display_name = 2;            // 显示名称
  string description = 3;             // 插件描述
  string version = 4;                 // 插件版本（语义化版本）
  string icon = 5;                    // 图标 URL 或 base64
  
  // 分类信息
  NodeCategory category = 6;          // 节点类别
  NodeType node_type = 7;             // 节点类型（触发器/处理器）
  repeated string tags = 8;           // 标签（用于搜索）
  
  // 参数定义
  repeated ParameterDef input_parameters = 10;   // 输入参数定义
  repeated ParameterDef output_parameters = 11;  // 输出参数定义
  
  // 凭证信息
  string credential_type = 15;        // 凭证类型（空表示不需要凭证）
  CredentialDef credential_def = 16;  // 凭证定义
  
  // 能力声明
  PluginCapabilities capabilities = 20;
  
  // 预留字段供未来扩展
  reserved 100 to 199;
}

// NodeCategory 节点类别
enum NodeCategory {
  CATEGORY_UNSPECIFIED = 0;
  CATEGORY_TRIGGER = 1;       // 触发器
  CATEGORY_ACTION = 2;        // 动作
  CATEGORY_CONDITION = 3;     // 条件判断
  CATEGORY_TRANSFORM = 4;     // 数据转换
  CATEGORY_INTEGRATION = 5;   // 集成（外部服务）
  CATEGORY_UTILITY = 6;       // 工具类
  CATEGORY_AI = 7;            // AI 相关
  CATEGORY_MEDIA = 8;         // 媒体处理
}

// NodeType 节点类型
enum NodeType {
  NODE_TYPE_UNSPECIFIED = 0;
  NODE_TYPE_TRIGGER = 1;      // 触发器（工作流入口）
  NODE_TYPE_PROCESSOR = 2;    // 处理器（普通节点）
  NODE_TYPE_BRANCH = 3;       // 分支节点
  NODE_TYPE_MERGE = 4;        // 合并节点
  NODE_TYPE_SUBFLOW = 5;      // 子工作流
}

// PluginCapabilities 插件能力声明
message PluginCapabilities {
  bool supports_streaming = 1;        // 支持流式输出
  bool supports_cancel = 2;           // 支持取消执行
  bool supports_retry = 3;            // 支持重试
  bool supports_batch = 4;            // 支持批量处理
  bool requires_credential = 5;       // 需要凭证
  int32 max_concurrent = 6;           // 最大并发数（0 表示无限制）
  int32 default_timeout_ms = 7;       // 默认超时时间
}

// ParameterDef 参数定义
message ParameterDef {
  string name = 1;                    // 参数名称
  string display_name = 2;            // 显示名称
  string description = 3;             // 参数描述
  ParameterType type = 4;             // 参数类型
  bool required = 5;                  // 是否必填
  Value default_value = 6;            // 默认值
  
  // 验证规则
  ParameterValidation validation = 10;
  
  // UI 相关
  string placeholder = 15;            // 占位符
  string hint = 16;                   // 提示信息
  ParameterUIType ui_type = 17;       // UI 展示类型
  repeated ParameterOption options = 18;  // 枚举选项（用于下拉框等）
  string group = 19;                  // 参数分组
  int32 order = 20;                   // 显示顺序
  
  // 条件显示
  string depends_on = 25;             // 依赖的参数名
  string show_when = 26;              // 显示条件表达式
}

// ParameterType 参数类型
enum ParameterType {
  PARAM_TYPE_UNSPECIFIED = 0;
  PARAM_TYPE_STRING = 1;
  PARAM_TYPE_INT = 2;
  PARAM_TYPE_FLOAT = 3;
  PARAM_TYPE_BOOL = 4;
  PARAM_TYPE_BYTES = 5;
  PARAM_TYPE_ARRAY = 6;
  PARAM_TYPE_OBJECT = 7;
  PARAM_TYPE_ENUM = 8;                // 枚举类型
  PARAM_TYPE_SECRET = 9;              // 敏感信息（密码等）
  PARAM_TYPE_EXPRESSION = 10;         // 表达式（支持变量引用）
  PARAM_TYPE_CODE = 11;               // 代码片段
  PARAM_TYPE_JSON = 12;               // JSON 数据
  PARAM_TYPE_FILE = 13;               // 文件
  PARAM_TYPE_URL = 14;                // URL
  PARAM_TYPE_DATETIME = 15;           // 日期时间
}

// ParameterUIType UI 展示类型
enum ParameterUIType {
  UI_TYPE_UNSPECIFIED = 0;
  UI_TYPE_TEXT = 1;                   // 单行文本
  UI_TYPE_TEXTAREA = 2;               // 多行文本
  UI_TYPE_NUMBER = 3;                 // 数字输入
  UI_TYPE_SWITCH = 4;                 // 开关
  UI_TYPE_SELECT = 5;                 // 下拉选择
  UI_TYPE_MULTI_SELECT = 6;           // 多选
  UI_TYPE_RADIO = 7;                  // 单选按钮组
  UI_TYPE_CHECKBOX = 8;               // 复选框
  UI_TYPE_DATE = 9;                   // 日期选择
  UI_TYPE_TIME = 10;                  // 时间选择
  UI_TYPE_DATETIME = 11;              // 日期时间选择
  UI_TYPE_COLOR = 12;                 // 颜色选择
  UI_TYPE_FILE = 13;                  // 文件上传
  UI_TYPE_CODE_EDITOR = 14;           // 代码编辑器
  UI_TYPE_JSON_EDITOR = 15;           // JSON 编辑器
  UI_TYPE_EXPRESSION = 16;            // 表达式编辑器
  UI_TYPE_KEY_VALUE = 17;             // 键值对编辑器
  UI_TYPE_SLIDER = 18;                // 滑动条
}

// ParameterOption 参数选项（用于枚举类型）
message ParameterOption {
  string value = 1;                   // 值
  string label = 2;                   // 显示标签
  string description = 3;             // 描述
  string icon = 4;                    // 图标
  bool disabled = 5;                  // 是否禁用
}

// ParameterValidation 参数验证规则
message ParameterValidation {
  // 字符串验证
  int32 min_length = 1;
  int32 max_length = 2;
  string pattern = 3;                 // 正则表达式
  
  // 数值验证
  double min_value = 10;
  double max_value = 11;
  
  // 数组验证
  int32 min_items = 20;
  int32 max_items = 21;
  
  // 通用
  repeated string allowed_values = 30;  // 允许的值列表
  string custom_validator = 31;         // 自定义验证器表达式
}

// CredentialDef 凭证定义
message CredentialDef {
  string type = 1;                    // 凭证类型
  string display_name = 2;            // 显示名称
  string description = 3;             // 描述
  repeated ParameterDef fields = 4;   // 凭证字段定义
  string auth_url = 5;                // OAuth 认证 URL（如果是 OAuth）
  string token_url = 6;               // OAuth Token URL
}

// =============================================================================
// Init - 初始化节点
// =============================================================================

message InitRequest {
  ExecutionContext context = 1;       // 执行上下文
  NodeConfig node_config = 2;         // 节点配置
  WorkflowConfig workflow_config = 3; // 工作流配置
  string server_endpoint = 4;         // 服务端 API 地址
  Credential credential = 5;          // 凭证信息
}

// NodeConfig 节点配置
message NodeConfig {
  string id = 1;                      // 节点 ID
  string name = 2;                    // 节点名称
  string kind = 3;                    // 节点类型（插件名称）
  map<string, Value> parameters = 4;  // 节点参数
  map<string, string> labels = 5;     // 标签
  NodePosition position = 6;          // 位置信息（用于可视化）
}

// NodePosition 节点位置
message NodePosition {
  double x = 1;
  double y = 2;
}

// WorkflowConfig 工作流配置
message WorkflowConfig {
  string id = 1;                      // 工作流 ID
  string name = 2;                    // 工作流名称
  string version = 3;                 // 工作流版本
  map<string, Value> global_vars = 4; // 全局变量
  map<string, string> env = 5;        // 环境变量
}

// Credential 凭证信息
message Credential {
  string type = 1;                    // 凭证类型
  map<string, Value> fields = 2;      // 凭证字段
  int64 expires_at_ms = 3;  // 过期时间（毫秒时间戳）
}

message InitResponse {
  bool success = 1;
  string error_code = 2;              // 错误码
  string error_message = 3;           // 错误信息
  map<string, string> metadata = 4;   // 初始化后的元数据
}

// =============================================================================
// Run - 执行节点
// =============================================================================

message RunRequest {
  ExecutionContext context = 1;       // 执行上下文
  map<string, Value> parameters = 2;  // 参数值
  map<string, Value> parent_output = 3;  // 父节点输出
  map<string, Value> global_vars = 4;    // 全局变量
  map<string, Value> local_vars = 5;     // 本地变量
}

message RunResponse {
  ResponseType type = 1;
  int64 timestamp_ms = 2;  // 时间戳（毫秒）
  
  oneof payload {
    LogPayload log = 10;              // 日志消息
    ProgressPayload progress = 11;    // 进度更新
    ResultPayload result = 12;        // 执行结果
    ErrorPayload error = 13;          // 错误信息
  }
}

// ResponseType 响应类型
enum ResponseType {
  RESPONSE_TYPE_UNSPECIFIED = 0;
  RESPONSE_TYPE_LOG = 1;              // 日志
  RESPONSE_TYPE_PROGRESS = 2;         // 进度
  RESPONSE_TYPE_RESULT = 3;           // 结果
  RESPONSE_TYPE_ERROR = 4;            // 错误
}

// LogPayload 日志消息
message LogPayload {
  LogLevel level = 1;
  string message = 2;
  map<string, string> fields = 3;     // 结构化日志字段
}

// LogLevel 日志级别
enum LogLevel {
  LOG_LEVEL_UNSPECIFIED = 0;
  LOG_LEVEL_TRACE = 1;
  LOG_LEVEL_DEBUG = 2;
  LOG_LEVEL_INFO = 3;
  LOG_LEVEL_WARN = 4;
  LOG_LEVEL_ERROR = 5;
}

// ProgressPayload 进度更新
message ProgressPayload {
  int32 current = 1;                  // 当前进度
  int32 total = 2;                    // 总进度
  string message = 3;                 // 进度消息
  double percentage = 4;              // 百分比 (0-100)
}

// ResultPayload 执行结果
message ResultPayload {
  map<string, Value> output = 1;      // 输出数据
  int32 branch_index = 2;             // 分支索引（用于条件分支）
  string next_node_id = 3;            // 下一个节点 ID（可选）
  ExecutionStatus status = 4;         // 执行状态
  int64 duration_ms = 5;              // 执行耗时（毫秒）
}

// ExecutionStatus 执行状态
enum ExecutionStatus {
  EXECUTION_STATUS_UNSPECIFIED = 0;
  EXECUTION_STATUS_SUCCESS = 1;       // 成功
  EXECUTION_STATUS_SKIPPED = 2;       // 跳过
  EXECUTION_STATUS_PARTIAL = 3;       // 部分成功
}

// ErrorPayload 错误信息
message ErrorPayload {
  string code = 1;                    // 错误码
  string message = 2;                 // 错误消息
  ErrorType error_type = 3;           // 错误类型
  bool retryable = 4;                 // 是否可重试
  int32 retry_after_ms = 5;           // 重试等待时间
  map<string, string> details = 6;    // 错误详情
  string stack_trace = 7;             // 堆栈跟踪（仅开发环境）
}

// ErrorType 错误类型
enum ErrorType {
  ERROR_TYPE_UNSPECIFIED = 0;
  ERROR_TYPE_VALIDATION = 1;          // 参数验证错误
  ERROR_TYPE_AUTHENTICATION = 2;      // 认证错误
  ERROR_TYPE_AUTHORIZATION = 3;       // 授权错误
  ERROR_TYPE_NOT_FOUND = 4;           // 资源不存在
  ERROR_TYPE_RATE_LIMIT = 5;          // 速率限制
  ERROR_TYPE_TIMEOUT = 6;             // 超时
  ERROR_TYPE_NETWORK = 7;             // 网络错误
  ERROR_TYPE_INTERNAL = 8;            // 内部错误
  ERROR_TYPE_EXTERNAL = 9;            // 外部服务错误
  ERROR_TYPE_CANCELLED = 10;          // 已取消
}

// =============================================================================
// Stop - 停止执行
// =============================================================================

message StopRequest {
  ExecutionContext context = 1;       // 执行上下文
  string reason = 2;                  // 停止原因
  bool force = 3;                     // 是否强制停止
}

message StopResponse {
  bool success = 1;
  string message = 2;
  StopStatus status = 3;
}

// StopStatus 停止状态
enum StopStatus {
  STOP_STATUS_UNSPECIFIED = 0;
  STOP_STATUS_STOPPED = 1;            // 已停止
  STOP_STATUS_STOPPING = 2;           // 正在停止
  STOP_STATUS_NOT_RUNNING = 3;        // 未在运行
  STOP_STATUS_CANNOT_STOP = 4;        // 无法停止
}

// =============================================================================
// TestCredential - 测试凭证
// =============================================================================

message TestCredentialRequest {
  string credential_type = 1;         // 凭证类型
  Credential credential = 2;          // 凭证信息
  int32 timeout_ms = 3;               // 超时时间
}

message TestCredentialResponse {
  bool success = 1;
  string error_code = 2;
  string error_message = 3;
  map<string, string> info = 4;       // 附加信息（如用户名、账户ID等）
}

// =============================================================================
// HealthCheck - 健康检查
// =============================================================================

message HealthCheckRequest {
  bool include_details = 1;           // 是否包含详细信息
}

message HealthCheckResponse {
  HealthStatus status = 1;            // 健康状态
  string message = 2;                 // 状态消息
  
  // 版本信息
  string plugin_version = 3;          // 插件版本
  string protocol_version = 4;        // 协议版本
  
  // 能力信息
  repeated string supported_features = 5;  // 支持的特性列表
  
  // 详细信息
  map<string, HealthCheckDetail> details = 6;  // 各组件健康详情
  
  // 资源使用
  ResourceUsage resource_usage = 7;   // 资源使用情况
  
  int64 checked_at_ms = 8;  // 检查时间（毫秒时间戳）
}

// HealthStatus 健康状态
enum HealthStatus {
  HEALTH_STATUS_UNSPECIFIED = 0;
  HEALTH_STATUS_HEALTHY = 1;          // 健康
  HEALTH_STATUS_DEGRADED = 2;         // 降级
  HEALTH_STATUS_UNHEALTHY = 3;        // 不健康
}

// HealthCheckDetail 健康检查详情
message HealthCheckDetail {
  HealthStatus status = 1;
  string message = 2;
  int64 latency_ms = 3;               // 延迟（毫秒）
}

// ResourceUsage 资源使用情况
message ResourceUsage {
  int64 memory_bytes = 1;             // 内存使用（字节）
  double cpu_percent = 2;             // CPU 使用率
  int32 goroutines = 3;               // Goroutine 数量
  int32 active_connections = 4;       // 活动连接数
}
